Prepare API models to support polls (read-only)

PIG208 opened the pull request:
Fixes: #165

gnprice commented on lib/api/model/model.dart:
```diff
@@ -570,6 +573,46 @@ enum MessageFlag {
 
+/// https://zulip.com/api/get-messages#response
+@JsonSerializable(fieldRename: FieldRename.snake)
+class Submessage {
```
Let's put this in the same file as the details of polls, and call it `submessage.dart`.
* That way it's a bit more general than just polls, and makes a natural home for todo lists when we add those in the future.
* Once we have this feature all working, I'd like to avoid thinking about submessages as much as possible. :slightly_smiling_face: And one generally can do that — they don't come up except when specifically working on polls. So it's helpful to have this type not be in between these core message classes that come up all over the Zulip app model.

gnprice reviewed (commented):
Cool! Looking forward to seeing the full version of this PR.

One high-level comment below, and one other:

Normally in our API code we just link to API docs and that covers the need for documentation. Here there aren't adequate API docs; so
* Please link instead to whatever sources you've made use of to work out what the API is — e.g. the web implementation, the zulip-mobile implementation, the server implementation.
* There are some doc comments on the API types in zulip-mobile. Those can I think be copied over mostly verbatim to describe the API semantics here — they're mostly playing the role that the docs at https://zulip.com/api/ normally play for us.

gnprice reviewed (commented):
Two other comments, from just taking a quick look today.

PIG208 commented:
I revamped the poll model and make it a part of the message object. Previously, the plan was to have a dedicated centralized store (like unreads) for polls, where we maintain a map keyed by the message ID. The old design has several downsides:
- `submessages` do not get further deserialized until it is processed by the dedicated store, while the intermediate `Submessage` objects are not useful.
- In contrast to `unreads` or `typingStatus`, polls have a one-to-one relationship to messages. Making them a property on `Message` simplifies the implementation.
- It just seems more appropriate to handle poll events inside `MessageStore`, so that we don't maintain methods like `handleMessageEvent` and `reconcileMessages` that would be needed for a separate store.

PIG208 commented:
The new approach will set `submessages` to null as soon as the constructor of `Message` is called. The constructor takes a look at the widget type and decides if the message contains a poll. If so, the converts the submessages into a singular `Poll` object and store that on the message.

PIG208 commented:
<details>

<summary>
UI Preview
</summary>

| screenshot | compare with spoiler |
| - | - |
| ![image](https://github.com/user-attachments/assets/f5ad728a-5204-430e-bb23-f9cf065c6b0d) | ![image](https://github.com/user-attachments/assets/e0ef3098-1fad-45ef-b98b-01d9caa67357) |

</details>

gnprice commented on lib/api/model/widget.dart:
```diff
@@ -0,0 +1,107 @@
+  final String text;
+  final Set<int> voters = {};
+
+  Iterable<String> getVoterNames(PerAccountStore store) =>
+    voters.map((userId) => store.users[userId]?.fullName).whereType<String>();
```
(just glancing through the changes from revision to revision, since this is still a draft)

In general when some list contains an unknown user, or an unknown stream/channel etc., we should still show that user (etc.), just with some fallback UI like the text "(unknown user)" to indicate that we don't know more details.

Otherwise we could get quite buggy-looking situations where we're showing the user misleading data — for example with a poll, we might say there were 3 votes but only list one name, or say there was just 1 vote but other users looking at the same poll see 3 votes, etc.

chrisbobbe commented:
Is this ready for my review? I see the title still says WIP.

PIG208 commented:
Yes. Updated the title. I plan to split out the UI commit (the last one) as it does not have tests. It is included here to make testing easier.

chrisbobbe commented on assets/l10n/app_en.arb:
```diff
@@ -483,5 +483,24 @@
   "@notifSelfUser": {
     "description": "Display name for the user themself, to show after replying in an Android notification"
+  },
+  "pollWidgetQuestionMissing": "(Untitled Poll)",
+  "@pollWidgetQuestionMissing": {
+    "description": "Text to display for a poll when the question is missing"
+  },
```
Not sure if this should be "(Untitled poll)" or something else; @alya might have thoughts on this.

PIG208 replied:
"(no title)" might be an option (inspired by "(no topic)")

chrisbobbe replied:
I think it's helpful to include "poll", so it's clear what kind of title-lacking thing we're rendering.

chrisbobbe reviewed (commented):
Thanks! Comments below, and could you please post screenshots of some polls, showing the various ways they can look (including all the added UI strings)?

PIG208 commented:
<details>

<summary>UI Preview</summary>

![Screenshot from 2024-07-29 15-54-04](https://github.com/user-attachments/assets/a1e10472-21b0-4e88-8613-9aba51970d0b)


</details>

PIG208 commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,269 @@
+
+sealed class SubmessageData {}
+
+/// The data encoded in a submessage to make the message a widget.
```
Yes, I was thinking of a better name that distinguishes these widgets from the Flutter widgets. Maybe we can use Zulip as a qualifying noun? I wanted to include Zulip in the class names as well, but went with this version without it for brevity.

chrisbobbe replied:
I think a note in this dartdoc just to clear up any confusion would be sufficient.

PIG208 commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,269 @@
+  @JsonKey(includeToJson: true)
+  WidgetType get widgetType => WidgetType.poll;
+
+  /// The initial question and options on the poll.
+  final PollWidgetExtraData extraData;
```
Looks like this does not happen to the linked message anymore. `{widget_type: poll, extra_data: {question: , options: []}}` is what I got. Maybe it got fixed at some point?

chrisbobbe replied:
Hmm, interesting. @gnprice, thoughts?

PIG208 replied:
For visibility on possibly related background: https://chat.zulip.org/#narrow/stream/101-design/topic/Old.20polls.20cause.20web.20exceptions.20and.20broken.20appearance

PIG208 replied:
I think [this fix](https://chat.zulip.org/#narrow/stream/101-design/topic/Old.20polls.20cause.20web.20exceptions.20and.20broken.20appearance/near/1814292) addresses the missing "extra_data" case, and ensures that "options" is always there ("question" predated "options" and was mandatory).

chrisbobbe replied:
Ah! Thanks for finding that discussion—I agree it seems like the malformed data on CZO has been cleaned up. And from that discussion, it also seems like we don't expect this kind of malformed data to be present on other servers either. It came from some experiments on CZO before the API was settled.

So I'm content to not handle a missing `extra_data` case.

PIG208 commented on lib/api/model/model.dart:
```diff
@@ -481,7 +483,9 @@ sealed class Message {
   @JsonKey(name: 'subject')
   final String topic;
-  // final List<string> submessages; // TODO handle
+  /// Poll data if "submessages" describe a poll, `null` otherwise.
+  @JsonKey(name: 'submessages', readValue: _readPoll, fromJson: _pollFromJson, toJson: _pollToJson)
+  Poll? poll;
```
That works for deserialization, but `toJson` yields unexpected results (`submessages` getting dumped to the "poll" field.) For supporting more widgets, the plan was to define a super class and rename `poll` to something generic like `zulipWidget`, that still reads from "submessages".

chrisbobbe replied:
I see; sure, that sounds fine.

PIG208 commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,269 @@
+/// As in [PollWidgetData.extraData].
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollWidgetExtraData {
+  final String question;
+  final List<String> options;
```
Yeah, those a good finds. Referencing https://github.com/zulip/zulip-flutter/pull/823#discussion_r1699090126, we can replicate the zulip mobile behavior by leaving `poll` as null as soon as the deserialization fails, so the message will be rendered into `ZulipContent`, instead of `PollContent`.

chrisbobbe replied:
I see; so we would keep `question` and `options` required—I agree with https://github.com/zulip/zulip-flutter/pull/823#discussion_r1699090126 that we can reasonably expect them to always be present—but also have error-handling, with a fallback to the usual message rendering style, in case of nasty surprises there (or anywhere)?

gnprice replied:
I think it's best to match the existing clients' behavior — that simplifies away any attempt to verify that unexpected shapes of data never happen. I've added a commit to do that:
afd3c9b7e api: Give defaults to question/options on PollWidgetExtraData

PIG208 commented on lib/model/message.dart:
```diff
@@ -267,4 +268,17 @@ class MessageStoreImpl with MessageStore {
+
+  void handleSubmessageEvent(SubmessageEvent event) {
+    final poll = messages[event.messageId]?.poll;
+    if (poll == null) return;
```
Yes. Probably we will want to do it earlier, at the end of `_readPoll`.

PIG208 commented on lib/model/message.dart:
```diff
@@ -267,4 +268,17 @@ class MessageStoreImpl with MessageStore {
+    if (poll == null) return;
+    poll.applyEvent(
+      event.senderId,
+      PollEvent.fromJson(event.content as Map<String, Object?>)
```
It would be challenging, because the expected data structure depends on the type of the widget, and that information comes separately on the Message object.

PIG208 replied:
This brings up a possible race. Will it be possible that this event arrives before we have the message?

If it does happen, we wouldn't be able to apply the event. I think the message when it later arrives ought to contain the up-to-date submessages, but I'm not entirely sure.

chrisbobbe replied:
I see. Yeah, the `SubmessageEvent` itself doesn't have a field saying it's for a poll, a todo list, or whatever. Yeah, I guess we'll leave this be.

PIG208 commented on lib/widgets/poll.dart:
```diff
@@ -0,0 +1,87 @@
+                    if (option.voters.isNotEmpty)
+                      Text(
+                        '(${option.getVoterNames(store, zulipLocalizations.unknownUserName).join(', ')})',
+                        style: const TextStyle(fontSize: 16, color: Color.fromARGB(255, 115, 115, 115)),
```
Perhaps something like this? (I'm updating the `zulipLocalizations.pollWidgetOptionsMissing` to use the grey color too with the secondary `TextStyle`):
```
    final pollTextStylePrimary = const TextStyle(fontSize: 16)
      .merge(weightVariableTextStyle(context, wght: 600));
    const pollTextStyleSecondary = TextStyle(
      fontSize: 16,
      color: Color.fromARGB(255, 115, 115, 115),
    );
```

chrisbobbe replied:
Sure, that seems reasonable.

Ah, and (not sure if I mentioned this already): how about putting the color on `ContentTheme`, so it's ready to go for #95 dark theme?

PIG208 replied:
Sounds good, will update that and let you know once it's ready for review.

PIG208 commented on test/api/model/submessage_test.dart:
```diff
@@ -55,4 +55,30 @@ void main() {
+      'vote': 'invalid',
+      'key': PollEvent.optionKey(senderId: null, optionIndex: 0)
+    })).isA<PollVoteEvent>().op.equals(VoteOp.unknown);
+  });
```
Yeah, this is likely helpful for testing the `PollEvent.toJson` logic.

PIG208 commented:
Updated the PR to address most of the review, except comments with some follow up questions.

chrisbobbe commented:
Looks like CI is failing; could you take a look?

PIG208 commented:
Should be fixed now. This was related to a generated code update.

chrisbobbe commented:
https://github.com/zulip/zulip-flutter/pull/823#discussion_r1700666328

> Sounds good, will update that and let you know once it's ready for review.

Cool; I'll await that from you.

PIG208 commented:
Updated the PR. Thanks for the review @chrisbobbe !

- Added `checkLogs` in separate commits. This refactor is not required, but it is helpful for verifying failure.
- Support dark mode. The dark mode colors are taken from the webapp.
- Added tests for PollWidget.
- Removed unused translation strings.
- Addressed review feedback.

PIG208 commented:
<details>

<summary>
UI Preview (dark mode)
</summary>

| Web | Flutter |
| - | - |
| ![image](https://github.com/user-attachments/assets/88f5b852-dad3-422c-9607-06d68918b547) | ![image](https://github.com/user-attachments/assets/dd551964-93c3-4cc6-a392-18bd3c6174ca) |

</details>

alya commented:
Cool!

- Another edge case to check is large numbers of votes (10+, or even 100+).
- Let's drop the `(` around "This poll has no options yet."; I think the italics are clear enough on their own.
- For "(no poll title)", let's try "No question.", ideally with bold + italics. If for some reason this is tricky, we don't have to do the italics here -- we try quite hard in the web UI to avoid polls without a question getting created in the first place.

PIG208 commented:
Thanks! Updated the test to handle the case with a lot of voters, missing question styling and the translation strings.

PIG208 commented:
<details>

<summary>
UI Preview (the only visual changes are the "No question." styling and the translation strings)
</summary>

![image](https://github.com/user-attachments/assets/4375d8ed-412d-4c45-9eaf-893c450524f8)

</details>

chrisbobbe commented on test/widgets/poll_test.dart:
```diff
@@ -0,0 +1,125 @@
+      ]
+    );
+
+    check(findInPoll(find.text('favorite letter')).evaluate()).single;
```
It would be nice if the test-setup code made it clearer why this is the string we want to look for. When I searched for this string in the file, I didn't get any results besides this one.

Could maybe:

- make `eg.pollWidgetDataJson` a function, and callers would pass the question they want. (So, searching 'favorite letter' in this file would identify that callsite.)
- instead of using the string literal 'favorite letter', use the value by reference, like `eg.pollWidgetDataJson['extra_data']['question']`
- rename `eg.pollWidgetDataJson` to something like `eg.pollWidgetDataJsonFavoriteLetter`, so the connection with the test-setup code is clearer, and also so the definition in example_data.dart gives a clue that its `question` property is a particular value that some tests will rely on

chrisbobbe reviewed (commented):
Thanks! Small comments below.

alya commented:
Ah, I was thinking about the visuals when there are a lot of votes. What does that look like? I remember Nehal doing some fiddling to get that to look OK in the web app.

PIG208 commented on test/widgets/poll_test.dart:
```diff
@@ -0,0 +1,125 @@
+    for (final (voter, optionIndex) in voterOptionIndexPairs) {
+      await store.handleEvent(eg.submessageEvent(123, voter.userId,
+        content: PollVoteEvent(
+          key: PollEvent.optionKey(senderId: null, optionIndex: optionIndex),
```
The `senderId` for `optionKey` is the ID of the user that created the option. I can see the ambiguity here, though.

PIG208 commented:
| Web | Flutter |
| - | - |
| ![Screenshot from 2024-08-06 15-36-14](https://github.com/user-attachments/assets/7bb2a51f-15e3-401e-8641-a00296c744b2) | ![Screenshot from 2024-08-06 15-36-56](https://github.com/user-attachments/assets/65ae36b9-58d7-4fb5-bd99-0e2d6cf8bc1e) |

PIG208 commented on test/test_log.dart:
```diff
@@ -0,0 +1,14 @@
+  assert(logHistory == null);
+  logHistory = [];
+  try {
+    predicate();
```
`FutureOr<T>` looks handy. Thanks!

PIG208 commented:
| light | dark |
| - | - |
| ![1000014152](https://github.com/user-attachments/assets/fd421203-d8ee-48cc-9bcb-c6e8226b02e2) | ![1000014153](https://github.com/user-attachments/assets/742dc4b0-d5df-4575-9248-7236ed33c7f5) |

- Updated both light and dark mode to have background color for the vote count.
- Updated dark mode to use the correct vote count text color.
- Addressed other review comments.

This should be ready for review

chrisbobbe commented:
Thanks! LGTM; over to Greg for his review.

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,285 @@
+  final Object? content;
+  final int messageId;
+  final int senderId;
+  final int id;
```
Does this get used? I think not — so in that case let's ignore it, like `messageId` above.

gnprice commented on lib/api/model/widget.dart:
```diff
@@ -0,0 +1,131 @@
+/// See also:
+/// - https://zulip.com/help/create-a-poll
+/// - https://github.com/zulip/zulip/blob/304d948416465c1a085122af5d752f03d6797003/web/shared/src/poll_data.ts
+class Poll {
```
Let's avoid having a file named `widget.dart` that isn't about Flutter widgets. The term "widget" is a highly salient one in our codebase, with a particular meaning; so it's confusing to have the term appear with a different meaning.

Where we have to use it (to refer to the Zulip API concept), we can manage that by making longer names that clearly namespace the reference into the context of the Zulip poll/todo "widget" concept. For this filename, though… how about just putting this all inside `submessage.dart`? It seems like most of that file is specific to polls anyway.

gnprice commented on lib/api/model/widget.dart:
```diff
@@ -0,0 +1,131 @@
+    required this.question,
+    required final List<String> options,
+  }) {
+    for (int index = 0; index < options.length; index += 1) {
+      // [null] is necessary for creating initial poll options.
+      _addOption(null, options[index], optionIndex: index);
```
I didn't understand what this comment was trying to say until a second or third reading. Now that I do understand it: it seems like it's not conveying any information beyond "yes, saying `null` here is correct". That's not a very helpful type of comment — given that the code says `null`, the reader pretty much already knows that the code's author thought saying `null` was correct.

A question that'd be more interesting to answer is *why* the code's author (you) thought that. After seeing more of the code, I think the answer is basically that [PollEvent.optionKey] says so. So something that'd be useful to say in a comment here is to point to that.

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,285 @@
+  factory PollEvent.fromJson(Map<String, Object?> json) {
+    final rawPollEventType = json['type'];
+    if (rawPollEventType == null) return UnknownPollEvent(json: json);
+    switch ($enumDecode(
```
similar comment to on WidgetType.fromJson above

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,285 @@
+
+/// A poll event when an option is added.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollOptionEvent extends PollEvent {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventType get type => PollEventType.newOption;
```
If the `type` field says "new option", let's say "new option" in our class name too. (Just like we do with `Event` subclasses.) That helps with correlating our logic with the API docs when they exist, or with other clients when those don't.

And in this case it makes for an intrinsically better name, too — if the semantics are (as the doc comment says) that this is *adding* an option, and not doing something else like removing or editing one, then it's good for the name to reflect that.

So:
```suggestion
class PollNewOptionEvent extends PollEvent {
  @override
  @JsonKey(includeToJson: true)
  PollEventType get type => PollEventType.newOption;
```

(modulo the other comment about the endings of these names)

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,285 @@
+
+/// As in [PollVoteEvent.op].
+@JsonEnum(valueField: 'apiValue')
+enum VoteOp {
```
Similar to the comment on `Option`:
```suggestion
enum PollVoteOp {
```

gnprice commented on lib/api/model/model.dart:
```diff
@@ -493,26 +500,72 @@ sealed class Message {
+        case UnsupportedWidgetData():
+          assert(debugLog('Unsupported widgetData: ${widgetData.json}'));
+      }
+    } on TypeError catch (e) {
+      assert(debugLog('Malformed data of widget type ${widgetData.widgetType}: $e\n${jsonEncode(submessages)}')); // TODO(log)
+    }
```
Why swallow this error? Seems like that'll lead to us having a `Message` object that we're unaware has a special meaning.

gnprice reviewed (commented):
Thanks @PIG208, and thanks @chrisbobbe and @alya for the previous reviews!

This is implementing a complex API, so I expect it to naturally take several more rounds of review. Here's a first round focused on the API (though I haven't read the entire API changes).

PIG208 commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,285 @@
+
+  @JsonKey(unknownEnumValue: SubmessageType.unknown)
+  final SubmessageType msgType;
+  /// [SubmessageData] encoded in JSON.
+  @JsonKey(readValue: readContent, toJson: contentToJson)
+  final Object? content;
```
Storing the parsed data back to `content` might be redundant, because these `Submessage` objects that are meant to be intermediate. I guess an alternative would be just storing the JSON string, and parse them when producing the `SubmessageData`.

gnprice replied:
> Storing the parsed data back to `content` might be redundant, because these `Submessage` objects that are meant to be intermediate.

Hmm, I don't really follow this. What would go wrong in putting the fully-parsed `SubmessageData` object here?

PIG208 replied:
Wasn't clear, sorry. We can't parse `SubmessageData` without looking at the content of the other submessages on the message. To parse it, we need to look at the index of the submessage, and the type of the Zulip widget (indicated by the content of the first submessage). These are something that can't be handled by `Submessage.fromJson` alone.

This kind of context aware parsing requirement makes it more appropriate to parse the `SubmessageData` on a method implemented by the `Message` class. So it is more of a restriction that we can't put fully-parsed `SubmessageData` there.

gnprice replied:
I see, annoying. (I think I've mentioned previously that this "submessages" API is not our best work.)

Let's have a comment explaining why `fromJson` can't fully decode this for us — the fact that it needs that context from outside. In particular I think the case where that constraint really bites is a SubmessageEvent, where we don't even know whether it's a poll widget or a todo widget until we look up the message ID in the message store.

And then yeah, I agree it may be cleaner to store just the JSON string here and let the consumer call `jsonDecode` as well as whatever "fromJson" sort of method it wants to call on the result of that. But if switching to that way seems like a pain, this way is OK too.

PIG208 replied:
Updated the PR to have a comment explaining that.

PIG208 commented on lib/api/model/widget.dart:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  /// See [Message._pollToJson].
+  final List<Submessage> submessages;
```
This exists purely for testing as mentioned in the commit message, specifically, for `prepareMessages` calls, where we serialize the `GetMessageResult` objects containing message objects with submessages. I think a way to get around this is refactoring out the part where `Message` objects are used in the test helpers, and modify the mappings directly, then returning empty lists should be fine.

gnprice replied:
I see. Yeah, that solution seems fine.

I think it would be a solvable problem to make `Message.toJson` produce an accurate `submessages`. But it would be kind of a normalized version — e.g. I think if someone voted, unvoted, then voted again, it'd record only the final vote because that's all the `Poll` object remembers. So for some tests that might not be adequate anyway.

PIG208 commented:
This update should have addressed most comments except the ones I commented on. Will get back to finishing off those later.

PIG208 commented:
Pushed again to remove preprocessing on `Submessage.content` and `SubmessageEvent.content`, and to add the workaround mentioned in https://github.com/zulip/zulip-flutter/pull/823#discussion_r1712436774.

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,397 @@
+            option.voters.add(senderId);
+          case PollVoteOp.remove:
+            option.voters.remove(senderId);
+          case PollVoteOp.unknown:
```
```suggestion
          case PollVoteOp.unknown: // TODO(log)
            return;
```

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,397 @@
+
+/// As in [PollEventSubmessage.type].
+@JsonEnum(fieldRename: FieldRename.snake)
+enum PollEventSubmessageType {
```
nit: let's put this right after PollEventSubmessage — it's closely tied to that base class (for example the `fromJson` there has a switch that's on exactly these values), and it provides the outline of what the subclasses will be.

gnprice reviewed (commented):
Thanks for the revision! Here's a complete review of the first few commits, describing the bulk of the API:
0d7a46425 api [nfc]: Avoid the use of dynamic on Message.
47cc9137a api: Add Submessage.
451dad9a7 submessage: Add SubmessageData classes for polls.

Those are close to merge now; the comments below are mostly small. I've inserted three additional commits for changes that were simpler to handle that way:
721e0d8c3 api [nfc]: Expand comments on Submessage; put fields in logical order
afd3c9b7e api: Give defaults to question/options on PollWidgetExtraData
9a975b49a api [nfc]: Expand comments on SubmessageData and subclasses

Then let's split up this PR, since the thread has gotten long. So after acting on the comments below, please drop the remaining commits from this PR and send them as a follow-up PR:
2043822a8 api: Construct polls data store from submessages.
de62dba0b event: Handle submessage event for polls.
acbd55774 (optional) test: Add a test helper for inspecting logs.
6b94ed394 poll: Support read-only poll widget UI.

PIG208 commented on lib/api/model/submessage.dart:
```diff
@@ -0,0 +1,397 @@
+  final String option;
+  /// The index of last [option] added by the sender.
+  @JsonKey(name: 'idx')
+  final int latestOptionIndex;
```
Also expanded this to `optionIndex` -> `idx` of `PollEventSubmessage.optionKey`.

PIG208 commented:
The PR has been updated. Split the second half of this PR to #885. Thanks for the review @gnprice!

gnprice reviewed (commented):
Thanks for the revision! One nit below, and then this will be ready to merge.

PIG208 commented:
Updated the PR.

gnprice commented:
Thanks! Looks good — merging.
